docs/mergeop/1_mergeop.md

```diff
@@ -0,0 +1,255 @@
+2. The output of MergeOp must be usable as an LLB result in the same way that outputs of SourceOp, FileOp and ExecOp are today. This allows MergeOp to work seamlessly with the rest of LLB.
+    1. For example, you must be able to use a MergeOp result as a mount in an ExecOp or as the base for a FileOp without needing to add special handling to those other Op types.
+
+## LLB
+
+In terms of the LLB go client, this design proposes a single new function in the client/llb package:
+* `Merge(states []State, opts ...MergeOpt) State`
+
+When merging states, the order matters, with "higher layer" states overriding and taking precedence over lower ones.
+* The order of arguments is lower->higher, which is mostly arbitrary in the same way big vs. little endian is, but was chosen because it’s most intuitive and allows you to add layers to the stack by appending to the slice.
+
+For now, Merge will only merge together the root (i.e. "/") of the provided states; there won’t be support for using selectors to merge subdirs of states.
+* While this would be technically possible to implement efficiently in Buildkit alone, problems arise when trying to export such states as images for use by runtimes that are unaware of MergeOp (i.e. containerd, docker, etc.) without resorting to copying data for every different state+selector combination.
+
+The opts will be used to configure the behavior when there are conflicts during the merge, which is detailed more in the next conflict detection doc.
+
+Some notes on forwards compatibility:
+* If we ever want to add the ability to use selectors on the input states of Merge, we can use the existing .Dir() method on States to implement choosing a subdir to merge. This means that for now we should return an error if users try to call Merge with a Dir value set to something besides "/"
+
+### Basic Example
+
+```
+stateA := Scratch().
+   	 File(Mkfile("/foo", 0777, []byte("A"))).
+   	 File(Mkfile("/a", 0777, []byte("A")))
+
+stateB := Scratch().
+   	 File(Mkfile("/foo", 0777, []byte("B"))).
+   	 File(Mkfile("/b", 0777, []byte("B")))
+
+mergedState := Merge(stateA, stateB) /*
+   	 /a   - contains A
+   	 /b   - contains B
+   	 /foo - contains B
+*/
+
+mergedState = Merge(stateB, stateA) /*
+   	 /a   - contains A
+   	 /b   - contains B
+   	 /foo - contains A
+*/
+```
+
+As you can see, the contents of /foo depend on the order of the merge, with the higher state taking precedence.
+
+/a and /b on the other hand are exclusive to stateA and stateB and thus get merged independently.
+
+### Deletion Example
+
+```
+stateA := Scratch().
+   	 File(Mkfile("/foo", 0777, []byte("A"))).
+   	 File(Mkfile("/a", 0777, []byte("A")))
+
+stateB := stateA.
+   	 File(Rm("/foo")).
+   	 File(Mkfile("/b", 0777, []byte("B")))
+
+stateC := Scratch().
+   	 File(Mkfile("/foo", 0777, []byte("C"))).
+   	 File(Mkfile("/c", 0777, []byte("C")))
```

Comment:
If here we do `File(Rm("/a"))`, will `Merge(stateB, stateC)` contain `/a` ? 
IIUC stateC doesn't contain `/a` so the whiteout won't be recorded by overlayfs and `Merge(stateB, stateC)` will do contain `/a`. Is this an expected result?

Reply:
Yeah good question but that's the expected behavior. In general I think it will be very rare for a user to want to create a layer with a deletion at a non-existent file, but in those obscure cases they can create a base layer with the file (i.e. `Scratch().File(Mkfile("/a"))`) and then do the deletion on top of that.

I don't love that but it's a fairly simple workaround that works and can be made seamless by higher-level tools consuming LLB if needed.